resources/assets/lib/notification-widget/with-mark-read.tsx

```diff
@@ -0,0 +1,94 @@
+/**
+ *    Copyright (c) ppy Pty Ltd <contact@ppy.sh>.
+ *
+ *    This file is part of osu!web. osu!web is distributed with the hope of
+ *    attracting more community contributions to the core ecosystem of osu!.
+ *
+ *    osu!web is free software: you can redistribute it and/or modify
+ *    it under the terms of the Affero GNU General Public License version 3
+ *    as published by the Free Software Foundation.
+ *
+ *    osu!web is distributed WITHOUT ANY WARRANTY; without even the implied
+ *    warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ *    See the GNU Affero General Public License for more details.
+ *
+ *    You should have received a copy of the GNU Affero General Public License
+ *    along with osu!web.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import * as _ from 'lodash';
+import * as React from 'react';
+import ItemProps from './item-props';
+
+export interface WithMarkReadProps {
+  canMarkRead: boolean;
+  markRead: () => void;
+  markReadFallback: (event: React.MouseEvent<HTMLElement>) => void;
+  markingAsRead: boolean;
+}
+
+interface State {
+  markingAsRead: boolean;
+}
+
+export function withMarkRead(Component: React.ComponentType<ItemProps & WithMarkReadProps>) {
```

Comment:
what the reason for this being a HOC instead of a component that renders `props.children` or using context?

Reply:
I don't think context works as each component has its own states and they're nested (`TypeGroup` &rarr; `ItemGroup` &rarr; `ItemCompact`).

Not sure how using `props.children` could work as it's currently used by two rather different type of components (`TypeGroup` and `Item*`).

Pretty sure this is the exact case for HOC (adding same logic between components) ?_?